client/modules/IDE/components/Editor/index.jsx

```diff
@@ -65,564 +73,599 @@ import * as PreferencesActions from '../../actions/preferences';
+    css(), // CSS support
+    keymap.of(standardKeymap),
+    Prec.highest(customKeymap), // Ensure custom keymap has the highest precedence
+    EditorView.updateListener.of((update) => {
+      if (update.docChanged) {
+        handleEditorChange(); // Handle document changes
+      }
+    }),
+    EditorView.updateListener.of((update) => {
+      if (update.domEvent && update.domEvent.type === 'keyup') {
+        handleKeyUp(update.domEvent);
+      }
+    }),
+    EditorView.lineWrapping,
+    hintExtension,
+    EditorView.decorations.compute([decorations], (state) => decorations)
+  ];
+
+  useEffect(() => {
+    if (!editorRef.current) return;
+
+    // if (!pickrRef.current) initColorPicker();
+
+    const startState = EditorState.create({
+      doc: file.content,
+      extensions: [...getCommonExtensions(), createThemeExtension(theme)]
+    });
+
+    const view = new EditorView({
+      state: startState,
+      parent: editorRef.current
+    });
+    viewRef.current = view;
+
+    const showHint = () => {
+      if (view) {
+        view.dispatch({
+          effects: autocompletion.startCompletion.of({
+            source: javascriptCompletion
+          })
+        });
       }
     };
-    const removeFocusedLinkElement = () => {
-      if (focusedLinkElement) {
-        focusedLinkElement.classList.remove('focused-hint-link');
-        focusedLinkElement.parentElement.parentElement.classList.remove(
-          'unfocused'
-        );
-        focusedLinkElement = null;
-        return true;
+
+    const onKeyDown = (e) => {
+      const mode = view?.state.facet(EditorState.languageData)?.[0];
+      if (/^[a-z]$/i.test(e.key) && (mode === 'css' || mode === 'javascript')) {
+        showHint();
+      }
+      if (e.key === 'Escape') {
+        e.preventDefault();
+        e.target.blur();
       }
-      return false;
     };
 
-    const hintOptions = {
-      _fontSize: this.props.fontSize,
-      completeSingle: false,
-      extraKeys: {
-        'Shift-Right': (cm, e) => {
-          const activeItemLink = document.querySelector(
-            `.CodeMirror-hint-active a`
-          );
-          if (activeItemLink) activeItemLink.click();
-        },
-        Right: (cm, e) => {
-          setFocusedLinkElement(true);
-        },
-        Left: (cm, e) => {
-          removeFocusedLinkElement();
-        },
-        Up: (cm, e) => {
-          const onLink = removeFocusedLinkElement();
-          e.moveFocus(-1);
-          setFocusedLinkElement(onLink);
-        },
-        Down: (cm, e) => {
-          const onLink = removeFocusedLinkElement();
-          e.moveFocus(1);
-          setFocusedLinkElement(onLink);
-        },
-        Enter: (cm, e) => {
-          if (focusedLinkElement) focusedLinkElement.click();
-          else e.pick();
-        }
-      },
-      closeOnUnfocus: false
+    view.dom.style.fontSize = `${fontSize}px`;
+
+    view.dom.addEventListener('keydown', onKeyDown);
```

Comment:
Any interaction with the DOM should be inside of a `useEffect`, where your cleanup function removes the previous listener when adding a new or updated one.  Otherwise you will be adding an additional event handler on every render.